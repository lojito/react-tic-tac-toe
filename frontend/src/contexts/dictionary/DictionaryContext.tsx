import { createContext } from "react";
import { Language } from "../../contexts/language/LanguageContext";
import data from "../../data/json/dictionary.json";

export type DictionaryEntry = Partial<
  Record<
    | "ABOUT"
    | "ABOUT_GAME_DEVELOPED_WITH"
    | "ABOUT_OTHER_HOME_PROJECTS"
    | "ABOUT_SOURCE_CODE"
    | "AUTH_FORM_EMAIL_LABEL"
    | "AUTH_FORM_EMAIL_PLACEHOLDER"
    | "AUTH_FORM_NAME_LABEL"
    | "AUTH_FORM_NAME_PLACEHOLDER"
    | "AUTH_FORM_PASSWORD_LABEL"
    | "AUTH_FORM_PASSWORD_PLACEHOLDER"
    | "AUTH_ERROR_SERVER_NOT_RESPONDING"
    | "AUTH_ERROR_VALIDATION_FAILED"
    | "AUTH_ERROR_UNKNOWN"
    | "BUTTON_BACK"
    | "BUTTON_DELETE_GAME"
    | "BUTTON_LOGOUT"
    | "BUTTON_PLAY"
    | "BUTTON_PLAY_AGAIN"
    | "BUTTON_SEE_GAME"
    | "BUTTON_SIGNIN"
    | "BUTTON_SIGNUP"
    | "CATEGORY"
    | "CATEGORY_ANIMALS"
    | "CATEGORY_FRUITS_AND_VEGETABLES"
    | "CATEGORY_GERMANY_LANDMARKS"
    | "CATEGORY_HAVANA_LANDMARKS"
    | "CATEGORY_MONTREAL_LANDMARKS"
    | "CATEGORY_PUPPIES"
    | "CATEGORY_SEINFELD"
    | "CATEGORY_SOCCER_PLAYERS"
    | "CATEGORY_SPAIN_LANDMARKS"
    | "CATEGORY_VANCOUVER_LANDMARKS"
    | "CONTACT"
    | "CONTACT_AUTHOR"
    | "CONTACT_EMAIL"
    | "CONTACT_GITHUB"
    | "CONTACT_LINKEDIN"
    | "FLAG_ENGLISH_HINT"
    | "FLAG_FRENCH_HINT"
    | "FLAG_SPANISH_HINT"
    | "GAME"
    | "GAME_OVER"
    | "GAME_STARTED_BY"
    | "GAME_STARTED_BY_COMPUTER"
    | "GAME_STARTED_BY_YOU"
    | "GAME_DELETING"
    | "GAME_DELETE_ERROR"
    | "GAME_SAVE_ERROR"
    | "GAME_SAVING"
    | "HISTORICAL"
    | "HISTORICAL_NO_DATA_YET"
    | "HISTORICAL_FETCHING_GAMES"
    | "HISTORICAL_FETCHING_GAME"
    | "HISTORICAL_FETCHING_GAMES_ERROR"
    | "HISTORICAL_FETCHING_GAME_ERROR"
    | "IMAGE_COMPUTER_LABEL"
    | "IMAGE_COMPUTER_TOOLTIP"
    | "IMAGE_REFRESH_TOOLTIP"
    | "IMAGE_USER_LABEL"
    | "IMAGE_USER_TOOLTIP"
    | "LOGIN"
    | "MESSAGE_DRAW"
    | "MESSAGE_LOST"
    | "MESSAGE_WON"
    | "PLAYING_LEVEL_EASY"
    | "PLAYING_LEVEL_LABEL"
    | "PLAYING_LEVEL_SMART"
    | "PLAYING_LEVEL_TOOLTIP"
    | "REPOSITORY"
    | "SIGNUP"
    | "WELCOME_BACK"
    | "START_GAME_LABEL"
    | "START_GAME_ANSWER_NO"
    | "START_GAME_TOOLTIP"
    | "START_GAME_ANSWER_YES",
    string
  >
>;

export type DictionaryContextValue = {
  dictionary: DictionaryEntry;
};

export const DictionaryContext = createContext<DictionaryContextValue>({
  dictionary: data[Language.English],
});
