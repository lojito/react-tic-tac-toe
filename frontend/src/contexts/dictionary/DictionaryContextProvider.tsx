import React, { FC, useContext } from "react";
import data from "../../data/json/dictionary.json";
import { LanguageContext } from "../language/LanguageContext";
import {
  DictionaryContext,
  DictionaryContextValue,
  DictionaryEntry,
} from "./DictionaryContext";

const DictionaryContextProvider: FC = ({ children }) => {
  const { language } = useContext(LanguageContext);

  const ctx: DictionaryContextValue = {
    dictionary: data[language] as DictionaryEntry,
  };

  return (
    <DictionaryContext.Provider value={ctx}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
