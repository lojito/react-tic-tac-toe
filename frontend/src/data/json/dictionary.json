[
  {
    "ABOUT": "About",
    "ABOUT_GAME_DEVELOPED_WITH": "Game developed with:",
    "ABOUT_OTHER_HOME_PROJECTS": "Other projects:",
    "ABOUT_SOURCE_CODE": "Source Code:",
    "AUTH_FORM_EMAIL_LABEL": "E-mail *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "E-mail",
    "AUTH_FORM_NAME_LABEL": "Name *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Name",
    "AUTH_FORM_PASSWORD_LABEL": "Password *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Password",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Server is not responding.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation failed. Invalid e-mail or password.",
    "AUTH_ERROR_UNKNOWN": "There was an error while trying to login or signup in the user.",
    "BUTTON_BACK": "Back",
    "BUTTON_DELETE_GAME": "Delete Game",
    "BUTTON_LOGOUT": "Logout",
    "BUTTON_PLAY": "Play Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Play again",
    "BUTTON_SEE_GAME": "See Game",
    "BUTTON_SIGNIN": "Log in",
    "BUTTON_SIGNUP": "Sign up",
    "CATEGORY": "Category",
    "CATEGORY_ANIMALS": "Animals",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits & Vegetables",
    "CATEGORY_GERMANY_LANDMARKS": "Germany landmarks",
    "CATEGORY_HAVANA_LANDMARKS": "Havana landmarks",
    "CATEGORY_MONTREAL_LANDMARKS": "Montreal landmarks",
    "CATEGORY_PUPPIES": "Puppies",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Soccer players",
    "CATEGORY_SPAIN_LANDMARKS": "Spain landmarks",
    "CATEGORY_VANCOUVER_LANDMARKS": "Vancouver landmarks",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Author:",
    "CONTACT_EMAIL": "E-mail:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "English",
    "FLAG_FRENCH_HINT": "French",
    "FLAG_SPANISH_HINT": "Spanish",
    "GAME": "Game",
    "GAME_OVER": "Game over",
    "GAME_STARTED_BY": "Started by",
    "GAME_STARTED_BY_COMPUTER": "Computer",
    "GAME_STARTED_BY_YOU": "You",
    "GAME_DELETING": "Removing the game data from the database...",
    "GAME_DELETE_ERROR": "An error ocurred while deleting the game from the database.",
    "GAME_SAVING": "Saving the game data...",
    "GAME_SAVE_ERROR": "An error ocurred while saving the game results into the database.",
    "HISTORICAL": "Historical",
    "HISTORICAL_NO_DATA_YET": "There is no historical data yet. Play some games and comeback.",
    "HISTORICAL_FETCHING_GAMES": "Fetching games...",
    "HISTORICAL_FETCHING_GAME": "Fetching game...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "An error ocurred while fetching the games from the database.",
    "HISTORICAL_FETCHING_GAME_ERROR": "An error ocurred while fetching the game from the database.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "The computer will play with this image.",
    "IMAGE_REFRESH_TOOLTIP": "Click to play with a different image.",
    "IMAGE_USER_LABEL": "You",
    "IMAGE_USER_TOOLTIP": "You will play with this image.",
    "LOGIN": "Log in",
    "MESSAGE_DRAW": "It's a draw!",
    "MESSAGE_LOST": "You lost!",
    "MESSAGE_WON": "You won!",
    "PLAYING_LEVEL_EASY": "Easy",
    "PLAYING_LEVEL_LABEL": "Level",
    "PLAYING_LEVEL_SMART": "Smart",
    "PLAYING_LEVEL_TOOLTIP": "How much thinking the computer does before playing.",
    "REPOSITORY": "Repository",
    "SIGNUP": "Sign up",
    "WELCOME_BACK": "Welcome back",
    "START_GAME_LABEL": "Start game",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Who starts playing the game: You or the computer?",
    "START_GAME_ANSWER_YES": "Yes"
  },
  {
    "ABOUT": "Acerca de",
    "ABOUT_GAME_DEVELOPED_WITH": "Juego desarrollado con:",
    "ABOUT_OTHER_HOME_PROJECTS": "Otros projectos:",
    "ABOUT_SOURCE_CODE": "Código fuente:",
    "AUTH_FORM_EMAIL_LABEL": "Correo electrónico *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Correo electrónico",
    "AUTH_FORM_NAME_LABEL": "Nombre *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nombre",
    "AUTH_FORM_PASSWORD_LABEL": "Contraseña *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Contraseña",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "El servidor no está respondiendo.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validación fallida. Correo electrónico o contraseña no válida.",
    "AUTH_ERROR_UNKNOWN": "Se produjo un error al intentar iniciar sesión o registrar al usuario.",
    "BUTTON_BACK": "Regresar",
    "BUTTON_DELETE_GAME": "Eliminar la partida",
    "BUTTON_LOGOUT": "Cerrar sesión",
    "BUTTON_PLAY": "Juega al Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jugar de nuevo",
    "BUTTON_SEE_GAME": "Ver la partida",
    "BUTTON_SIGNIN": "Iniciar sesión",
    "BUTTON_SIGNUP": "Inscribirse",
    "CATEGORY": "Categoría",
    "CATEGORY_ANIMALS": "Animales",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Frutas y vegetales",
    "CATEGORY_GERMANY_LANDMARKS": "Monumentos de Alemania",
    "CATEGORY_HAVANA_LANDMARKS": "Monumentos de La Habana",
    "CATEGORY_MONTREAL_LANDMARKS": "Monumentos de Montreal",
    "CATEGORY_PUPPIES": "Cachorros",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Jugadores de fútbol",
    "CATEGORY_SPAIN_LANDMARKS": "Monumentos de España",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monumentos de Vancouver",
    "CONTACT": "Contacto",
    "CONTACT_AUTHOR": "Autor:",
    "CONTACT_EMAIL": "Correo electrónico:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Inglés",
    "FLAG_FRENCH_HINT": "Francés",
    "FLAG_SPANISH_HINT": "Español",
    "GAME": "Juego",
    "GAME_OVER": "Partida acabada",
    "GAME_STARTED_BY": "Comenzada por",
    "GAME_STARTED_BY_COMPUTER": "Ordenador",
    "GAME_STARTED_BY_YOU": "Ti",
    "GAME_DELETING": "Eliminando el juego de la base de datos...",
    "GAME_DELETE_ERROR": "Ocurrió un error al eliminar el juego de la base de datos.",
    "GAME_SAVING": "Guardando los datos del juego...",
    "GAME_SAVE_ERROR": "Ocurrió un error al salvar los resultados del juego de la base de datos.",
    "HISTORICAL": "Histórico",
    "HISTORICAL_NO_DATA_YET": "No hay historial aún. Juega algunas partidas y regresa más tarde.",
    "HISTORICAL_FETCHING_GAMES": "Obteniendo juegos...",
    "HISTORICAL_FETCHING_GAME": "Obteniendo juego...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Ocurrió un error al recuperar los juegos de la base de datos.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Ocurrió un error al recuperar el juego de la base de datos.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "El ordenador jugará con esta imagen.",
    "IMAGE_REFRESH_TOOLTIP": "Haz click para jugar con una imagen diferente.",
    "IMAGE_USER_LABEL": "Tú",
    "IMAGE_USER_TOOLTIP": "Tú jugaras con esta imagen.",
    "LOGIN": "Iniciar sesión",
    "MESSAGE_DRAW": "Empate!",
    "MESSAGE_LOST": "Perdiste!",
    "MESSAGE_WON": "Ganaste!",
    "PLAYING_LEVEL_EASY": "Fácil",
    "PLAYING_LEVEL_LABEL": "Nivel de juego",
    "PLAYING_LEVEL_SMART": "Difícil",
    "PLAYING_LEVEL_TOOLTIP": "Cuánto tiempo el ordenador piensa antes de jugar.",
    "REPOSITORY": "Banco",
    "SIGNUP": "Regístrate",
    "WELCOME_BACK": "Bienvenido de nuevo",
    "START_GAME_LABEL": "Tú empiezas",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Quién empieza la partida: Tú o el ordenador?",
    "START_GAME_ANSWER_YES": "Sí"
  },
  {
    "ABOUT": "À propos de",
    "ABOUT_GAME_DEVELOPED_WITH": "Jeu développé avec:",
    "ABOUT_OTHER_HOME_PROJECTS": "D'autres projects:",
    "ABOUT_SOURCE_CODE": "Code source:",
    "AUTH_FORM_EMAIL_LABEL": "Courrier électronique *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Courrier électronique",
    "AUTH_FORM_NAME_LABEL": "Nom *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nom",
    "AUTH_FORM_PASSWORD_LABEL": "Mot de passe *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Mot de passe",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Le serveur ne répond pas.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation échouée. Email ou mot de passe invalide.",
    "AUTH_ERROR_UNKNOWN": "Une erreur s'est produite lors de la tentative de connexion ou d'inscription de l'utilisateur.",
    "BUTTON_BACK": "Retour",
    "BUTTON_DELETE_GAME": "Retirer la partie",
    "BUTTON_LOGOUT": "Se déconnecter",
    "BUTTON_PLAY": "Joue au Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jouer à nouveau",
    "BUTTON_SEE_GAME": "Voir la partie",
    "BUTTON_SIGNIN": "Se connecter",
    "BUTTON_SIGNUP": "S'inscrire",
    "CATEGORY": "Catégorie",
    "CATEGORY_ANIMALS": "Des animaux",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits et légumes",
    "CATEGORY_GERMANY_LANDMARKS": "Monuments L'Allemagne",
    "CATEGORY_HAVANA_LANDMARKS": "Monuments de La Havane",
    "CATEGORY_MONTREAL_LANDMARKS": "Monuments de Montréal",
    "CATEGORY_PUPPIES": "Des chiots",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Joueurs du soccer",
    "CATEGORY_SPAIN_LANDMARKS": "Monuments L'Espagne",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monuments de Vancouver",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Auteur:",
    "CONTACT_EMAIL": "Courrier électronique:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Anglais",
    "FLAG_FRENCH_HINT": "Français",
    "FLAG_SPANISH_HINT": "Espagnol",
    "GAME": "Jeu",
    "GAME_OVER": "La partie est finie",
    "GAME_STARTED_BY": "Commencé par",
    "GAME_STARTED_BY_COMPUTER": "Ordinateur",
    "GAME_STARTED_BY_YOU": "Toi",
    "GAME_DELETING": "En train de supprimer le jeu de la bsae de données...",
    "GAME_DELETE_ERROR": "Une erreur s'est produite lors de la suppression du jeu de la base de données.",
    "GAME_SAVING": "En train de sauvegarder les données du jeu...",
    "GAME_SAVE_ERROR": "Une erreur s'est produite lors de l'enregistrement des résultats du jeu dans la base de données.",
    "HISTORICAL": "Historique",
    "HISTORICAL_NO_DATA_YET": "Il n'y a pas encore d'historique. Reviens un peu plus tard après avoir joué quelques parties.",
    "HISTORICAL_FETCHING_GAMES": "Récupération des jeux...",
    "HISTORICAL_FETCHING_GAME": "Récupération du jeu...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Une erreur s'est produite lors de la récupération des jeux de la base de données.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Une erreur s'est produite lors de la récupération du jeu de la base de données.",
    "IMAGE_COMPUTER_LABEL": "Ordi",
    "IMAGE_COMPUTER_TOOLTIP": "L'ordinateur va jouer avec cette image.",
    "IMAGE_REFRESH_TOOLTIP": "Fais clic à fin de jouer avec une autre image.",
    "IMAGE_USER_LABEL": "Toi",
    "IMAGE_USER_TOOLTIP": "Tu vas jouer avec cette image.",
    "LOGIN": "Se connecter",
    "MESSAGE_DRAW": "Match nul!",
    "MESSAGE_LOST": "Tu as perdu!",
    "MESSAGE_WON": "Tu as gagné!",
    "PLAYING_LEVEL_EASY": "Facile",
    "PLAYING_LEVEL_LABEL": "Niveau de jeu",
    "PLAYING_LEVEL_SMART": "Difficile",
    "PLAYING_LEVEL_TOOLTIP": "Quelle est la réflexion de l'ordinateur avant de jouer.",
    "REPOSITORY": "Banque",
    "SIGNUP": "S'inscrire",
    "WELCOME_BACK": "Ravis de vous retrouver",
    "START_GAME_LABEL": "Premier",
    "START_GAME_ANSWER_NO": "Non",
    "START_GAME_TOOLTIP": "Qui commence la partie: L'ordinateur ou toi?",
    "START_GAME_ANSWER_YES": "Oui"
  }
]
