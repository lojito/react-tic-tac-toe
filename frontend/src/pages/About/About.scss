.about {
  width: 95%;
  margin: 10px auto 0;

  ul {
    margin-top: 0;

    li {
      margin-top: 5px;
    }
  }

  span {
    font-weight: bold;
  }
}
