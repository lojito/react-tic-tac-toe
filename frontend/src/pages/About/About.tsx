import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import "./About.scss";

const About: FC = () => {
  const {
    dictionary: {
      ABOUT_GAME_DEVELOPED_WITH,
      ABOUT_SOURCE_CODE,
      ABOUT_OTHER_HOME_PROJECTS,
    },
  } = useContext(DictionaryContext);

  const techStack = [
    { name: "Node.js", url: "https://nodejs.org" },
    { name: "Express", url: "https://expressjs.com/" },
    {
      name: "Mongoose",
      url: "https://mongoosejs.com/docs/api.html",
    },
    {
      name: "MongoDB",
      url: "https://www.mongodb.com/docs/manual/introduction/",
    },
    { name: "Typescript", url: "https://www.typescriptlang.org/" },
    { name: "React.js", url: "https://reactjs.org/" },
    { name: "React Hooks", url: "https://reactjs.org/docs/hooks-intro.html" },
    { name: "Context API", url: "https://reactjs.org/docs/context.html" },
    {
      name: "React Testing Library",
      url: "https://testing-library.com/docs/react-testing-library/intro/",
    },
    { name: "Jest", url: "https://jestjs.io/" },
    {
      name: "HTML5",
      url: "https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5",
    },
    { name: "CSS3", url: "https://www.w3.org/TR/selectors-3/" },
    { name: "SASS", url: "https://sass-lang.com/" },
    {
      name: "REST API",
      url: "https://www.tutorialspoint.com/restful/index.htm/",
    },
    { name: "GraphQL", url: "https://graphql.org/" },
  ];

  const techStackLi = techStack.map(({ name, url }) => {
    return (
      <li key={name}>
        <a target="_blank" href={url} rel="noopener noreferrer">
          {name}
        </a>
      </li>
    );
  });

  const source = "https://github.com/lojito/tic-tac-toe";

  const otherProjects = [
    {
      name: "Puzzle (Angular)",
      url: "https://github.com/lojito/angular-puzzle",
    },
    {
      name: "Matching pairs (Vue.js)",
      url: "https://github.com/lojito/matching-pairs",
    },
    {
      name: "Tic-tac-toe 4x4 (Django)",
      url: "https://github.com/lojito/django-tic-tac-toe",
    },
    {
      name: "Tic-tac-toe (Python)",
      url: "https://github.com/lojito/python-tic-tac-toe",
    },
    {
      name: "Base converter (Rust)",
      url: "https://github.com/lojito/baseconverter",
    },
    {
      name: "In-memory database engine (C)",
      url: "https://github.com/lojito/c_champions_league",
    },
    {
      name: "Bubble sort sorting algorithm (Assembly)",
      url: "https://github.com/lojito/assembly-bubble-sort",
    },
  ];

  const otherProjectsLi = otherProjects.map(({ name, url }) => {
    return (
      <li key={name}>
        <a target="_blank" href={url} rel="noopener noreferrer">
          {name}
        </a>
      </li>
    );
  });

  return (
    <div className="about" data-testid="about">
      <div>
        <span>{ABOUT_GAME_DEVELOPED_WITH}</span>
        <ul>{techStackLi}</ul>
      </div>
      <div>
        <span>{ABOUT_SOURCE_CODE}</span>
        <ul>
          <li>
            <a target="_blank" href={source} rel="noopener noreferrer">
              {source}
            </a>
          </li>
        </ul>
      </div>
      <span>{ABOUT_OTHER_HOME_PROJECTS}</span>
      <ul>{otherProjectsLi}</ul>
    </div>
  );
};

export default About;
