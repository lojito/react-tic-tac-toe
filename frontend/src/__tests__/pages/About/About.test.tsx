import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import { DictionaryContext } from "../../../contexts/dictionary/DictionaryContext";
import About from "../../../pages/About/About";

describe("About", () => {
  const mockDictionary = {
    ABOUT_GAME_DEVELOPED_WITH: "This game has been developed with:",
    ABOUT_SOURCE_CODE: "Source Code:",
    ABOUT_OTHER_HOME_PROJECTS: "Other projects:",
  };

  const renderAbout = () => {
    const utils = render(
      <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
        <About />
      </DictionaryContext.Provider>
    );
    const aboutWrapper = screen.getByTestId("about");
    const gameDevelopedWith = screen.getByText(
      mockDictionary.ABOUT_GAME_DEVELOPED_WITH
    );
    const sourceCode = screen.getByText(mockDictionary.ABOUT_SOURCE_CODE);
    const otherHomeProjects = screen.getByText(
      mockDictionary.ABOUT_OTHER_HOME_PROJECTS
    );

    return {
      ...utils,
      aboutWrapper,
      gameDevelopedWith,
      sourceCode,
      otherHomeProjects,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderAbout();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the About info and apply the 'about' class", () => {
    const { aboutWrapper } = renderAbout();

    expect(aboutWrapper).toBeInTheDocument();
    expect(aboutWrapper).toHaveClass("about");
  });

  it("should display the texts 'Game developed with', 'Source Code', 'Other projects'", () => {
    const { gameDevelopedWith, sourceCode, otherHomeProjects } = renderAbout();

    expect(gameDevelopedWith).toBeInTheDocument();
    expect(sourceCode).toBeInTheDocument();
    expect(otherHomeProjects).toBeInTheDocument();
  });

  it("should display the teck stack list", () => {
    const techStack = [
      "Node.js",
      "Express",
      "Mongoose",
      "MongoDB",
      "Typescript",
      "React.js",
      "React Hooks",
      "Context API",
      "React Testing Library",
      "Jest",
      "HTML5",
      "CSS3",
      "SASS",
    ];

    renderAbout();

    for (const tech of techStack) {
      expect(screen.getByRole("link", { name: tech })).toBeInTheDocument();
    }
  });

  it("should display the 'Source code' info", () => {
    renderAbout();

    expect(
      screen.getByRole("link", { name: /tic-tac-toe/ })
    ).toBeInTheDocument();
  });

  it("should display the projects list", () => {
    const otherProjects = [
      "Puzzle (Angular)",
      "Matching pairs (Vue.js)",
      "Tic-tac-toe 4x4 (Django)",
      "Tic-tac-toe (Python)",
      "Base converter (Rust)",
      "In-memory database engine (C)",
      "Bubble sort sorting algorithm (Assembly)",
    ];

    renderAbout();

    for (const project of otherProjects) {
      expect(screen.getByRole("link", { name: project })).toBeInTheDocument();
    }
  });
});
