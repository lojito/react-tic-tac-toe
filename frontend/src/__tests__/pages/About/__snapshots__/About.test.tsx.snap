// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About should render correctly 1`] = `
<DocumentFragment>
  <div
    class="about"
    data-testid="about"
  >
    <div>
      <span>
        This game has been developed with:
      </span>
      <ul>
        <li>
          <a
            href="https://nodejs.org"
            rel="noopener noreferrer"
            target="_blank"
          >
            Node.js
          </a>
        </li>
        <li>
          <a
            href="https://expressjs.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Express
          </a>
        </li>
        <li>
          <a
            href="https://mongoosejs.com/docs/api.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            Mongoose
          </a>
        </li>
        <li>
          <a
            href="https://www.mongodb.com/docs/manual/introduction/"
            rel="noopener noreferrer"
            target="_blank"
          >
            MongoDB
          </a>
        </li>
        <li>
          <a
            href="https://www.typescriptlang.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Typescript
          </a>
        </li>
        <li>
          <a
            href="https://reactjs.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            React.js
          </a>
        </li>
        <li>
          <a
            href="https://reactjs.org/docs/hooks-intro.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            React Hooks
          </a>
        </li>
        <li>
          <a
            href="https://reactjs.org/docs/context.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            Context API
          </a>
        </li>
        <li>
          <a
            href="https://testing-library.com/docs/react-testing-library/intro/"
            rel="noopener noreferrer"
            target="_blank"
          >
            React Testing Library
          </a>
        </li>
        <li>
          <a
            href="https://jestjs.io/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Jest
          </a>
        </li>
        <li>
          <a
            href="https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5"
            rel="noopener noreferrer"
            target="_blank"
          >
            HTML5
          </a>
        </li>
        <li>
          <a
            href="https://www.w3.org/TR/selectors-3/"
            rel="noopener noreferrer"
            target="_blank"
          >
            CSS3
          </a>
        </li>
        <li>
          <a
            href="https://sass-lang.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            SASS
          </a>
        </li>
        <li>
          <a
            href="https://www.tutorialspoint.com/restful/index.htm/"
            rel="noopener noreferrer"
            target="_blank"
          >
            REST API
          </a>
        </li>
        <li>
          <a
            href="https://graphql.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            GraphQL
          </a>
        </li>
      </ul>
    </div>
    <div>
      <span>
        Source Code:
      </span>
      <ul>
        <li>
          <a
            href="https://github.com/lojito/tic-tac-toe"
            rel="noopener noreferrer"
            target="_blank"
          >
            https://github.com/lojito/tic-tac-toe
          </a>
        </li>
      </ul>
    </div>
    <span>
      Other projects:
    </span>
    <ul>
      <li>
        <a
          href="https://github.com/lojito/angular-puzzle"
          rel="noopener noreferrer"
          target="_blank"
        >
          Puzzle (Angular)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/matching-pairs"
          rel="noopener noreferrer"
          target="_blank"
        >
          Matching pairs (Vue.js)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/django-tic-tac-toe"
          rel="noopener noreferrer"
          target="_blank"
        >
          Tic-tac-toe 4x4 (Django)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/python-tic-tac-toe"
          rel="noopener noreferrer"
          target="_blank"
        >
          Tic-tac-toe (Python)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/baseconverter"
          rel="noopener noreferrer"
          target="_blank"
        >
          Base converter (Rust)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/c_champions_league"
          rel="noopener noreferrer"
          target="_blank"
        >
          In-memory database engine (C)
        </a>
      </li>
      <li>
        <a
          href="https://github.com/lojito/assembly-bubble-sort"
          rel="noopener noreferrer"
          target="_blank"
        >
          Bubble sort sorting algorithm (Assembly)
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
